metagamer/environments/qagent.py

```diff
@@ -1,144 +1,432 @@
+        )
+
+    def prune_qdict(self, state: str, valid_actions: Sequence[int]):
+        """Optional function that allows us to reduce space"""
+        if set(valid_actions) - self.qdict[state].keys():
+            raise ValueError("We have found valid actions not in our qtable for state")
+        elif self.qdict[state].keys() - set(valid_actions):
+            # we can prune down since we don't need the other moves
+            self.qdict[state] = {k: self.qdict[state][k] for k in valid_actions}
+
+    def get_max(self, state: str, valid_actions: Optional[Sequence[int]] = None):
+        """get maximum q score for a state over actions"""
+        if valid_actions:
+            self.prune_qdict(state, valid_actions)
+        return max(self.qdict[state].values())
+
+    def get_arg_max(self, state, valid_actions: Optional[Sequence[int]] = None):
+        """
+        which action gives max q score. If identical max q scrores exist,
+           this'll choose somehing at random
+        """
+        mx = self.get_max(state, valid_actions)
+        # get all moves that return optimal outcome
+        all_arg_maxes = [k for k, v in self.qdict[state].items() if v == mx]
+        return random.choice(all_arg_maxes)
+
 
-        test_window = int(num_epsiodes / 20)
+class TicTacToeRunner:
+    def __init__(self, agent1: Agent, agent2: Agent):
+        self.env = tictactoe.TicTacToeEnv()
+        # player order
+        self.agent1 = agent1
+        self.agent2 = agent2
 
-        num_steps = np.zeros(num_epsiodes)
-        eps_vals = np.zeros(num_epsiodes)
+        # interact with the same environment, but through their preferred representation
+        self.agent1_env = self.agent1.wrapper(self.env)
+        # player 2 wrapper multiplies reward by -1
+        self.agent2_env = self.agent2.wrapper(self.env)
+
+    def train(self, num_epsiodes):
+        """"""
+
+        results = np.zeros(num_epsiodes)
 
         for i in range(num_epsiodes):
-            state = self.env.reset()
-            steps = 0
-            epsilon -= epsdecay
+            self.env.reset()
 
-            done = False
+            # run first two moves
 
-            while not done:
-                action = self.get_action(state, epsilon=epsilon)
+            p1_state = self.agent1_env.get_observation()
+            p1_action = self.agent1.get_action(p1_state, self.agent1_env.valid_actions)
+            _, p1_reward, done, info = self.agent1_env.step(p1_action, 1)
 
-                new_state, reward, done, info = self.env.step(action, self.player)
-                if i == num_epsiodes - 1:
-                    self.env.render()
+            while True:
+
+                p2_state = self.agent2_env.get_observation()
+                p2_action = self.agent2.get_action(
+                    p2_state, self.agent2_env.valid_actions
+                )
+                _, reward, done, info = self.agent2_env.step(p2_action, -1)
```

Comment:
can't recall what this `_,` implies, though I think we discussed. Just for an unneeded argument?

Reply:
yeah, in this case the observation returned is not useful since we want it in env1's representation so we drop it